nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/lucene/DocsReader.java

```diff
@@ -100,101 +96,61 @@ private ProvenanceEventRecord getRecord(final Document d, final RecordReader rea
-                        if (potentialFiles.isEmpty()) {
-                            logger.warn("Could not find Provenance Log File with basename {} in the "
-                                    + "Provenance Repository; assuming file has expired and continuing without it", storageFilename);
-                            storageFilesToSkip.add(storageFilename);
-                            continue;
-                        }
+        Map<String, List<Document>> byStorageNameDocGroups = LuceneUtil.groupDocsByStorageFileName(docs);
 
-                        if (potentialFiles.size() > 1) {
-                            throw new FileNotFoundException("Found multiple Provenance Log Files with basename " +
-                                    storageFilename + " in the Provenance Repository");
-                        }
+        int eventsReadThisFile = 0;
+        int logFileCount = 0;
 
-                        for (final File file : potentialFiles) {
-                            try {
-                                if (reader != null) {
-                                    logger.debug("Read {} records from previous file", eventsReadThisFile);
-                                }
-
-                                reader = RecordReaders.newRecordReader(file, allProvenanceLogFiles, maxAttributeChars);
-                                matchingRecords.add(getRecord(d, reader));
-                                eventsReadThisFile = 1;
-
-                                if ( retrievalCount.incrementAndGet() >= maxResults ) {
-                                    break;
-                                }
-                            } catch (final IOException e) {
-                                throw new IOException("Failed to retrieve record " + d + " from Provenance File " + file + " due to " + e, e);
+        for (String storageFileName : byStorageNameDocGroups.keySet()) {
+            File provenanceEventFile = LuceneUtil.getProvenanceLogFile(storageFileName, allProvenanceLogFiles);
+            if (provenanceEventFile != null) {
+                try (RecordReader reader = RecordReaders.newRecordReader(provenanceEventFile, allProvenanceLogFiles,
+                        maxAttributeChars)) {
+                    for (Document document : byStorageNameDocGroups.get(storageFileName)) {
+                        ProvenanceEventRecord eRec = this.getRecord(document, reader);
+                        if (eRec != null) {
+                            matchingRecords.add(eRec);
+                            eventsReadThisFile++;
+
+                            if (retrievalCount.incrementAndGet() >= maxResults) {
+                                break;
                             }
                         }
                     }
-                } finally {
-                    lastStorageFilename = storageFilename;
+                } catch (Exception e) {
```

Comment:
Did you consider breaking out the different types exceptions here, which could help with diagnosis? Also, although unlikely (and possibly not plausible), the reader could throw an exception on close. Did you consider a closeQuietly instead of a try-with-resources?